### PR TITLE
Manufacturable deployable warp beacons

### PR DESCRIPTION
**Pull request by Kubius**
[FEAT]
## About the PR 

Introduces a manufacturable "beacon kit" that can be turned into a deployable warp beacon, and modifies the base warp beacon a bit to allow pack-up of deployed beacons.

Some notes about the implementation:

- The beacon kit uses all three basic materials (currently 10 of each), and must be further assembled after it's manufactured, requiring four rods, a unit of wire, and a soldering iron to perform the final connections.
- Beacons can be given a custom name using a multitool, with a length limit of 30 characters. It will always be prefixed with "Buoy" to distinguish player-made beacons from pre-existing ones.
- Deploying a beacon (using a wrench) requires a 5x5 area of vehicle-pathable turfs (centered on the beacon), and doesn't work on restricted Z-levels.
- Warp beacons now have a "packable" variable that distinguishes the deployable ones from fixed installations; fixed installations can't be renamed or packed for redeployment.

I originally coded this in the 2016 release, but I've fiddled with the code a bit, including updating it with a custom action bar that handles the assembly stages - assembly and operation appears to function solidly ingame.

## Why's this needed? 

Allows for convenient return to an off-station location (big help for construction projects and potentially mining as well), or replacement of a beacon that was destroyed.

## Changelog

```
(u)Kubius:
(*)Deployable warp beacons can now be manufactured in ship component fabricators. Some assembly required.
```

### CODE REVIEW

File: code/modules/transport/pods/warp_travel.dm [hunk 1]
```diff
@@ -144,4 +173,186 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+		if (duration_i)
+			duration = duration_i
+		if (ishuman(owner))
+			var/mob/living/carbon/human/H = owner
+			if (H.mind.assigned_role == "Chief Engineer")
+				duration = round(duration / 2)
+			if (H.mind.assigned_role == "Engineer")
+				duration = round(duration / 2)
+			if (H.mind.assigned_role == "Mechanic")
+				duration = round(duration / 2)
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:296]: engineering training trait perhaps?

Comment (Kubius) [reply code/modules/transport/pods/warp_travel.dm]: Yep. 2dadc92

File: code/modules/transport/pods/warp_travel.dm [hunk 2]
```diff
@@ -144,4 +173,182 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
 		playsound(src.loc, "warp", 50, 1, 0.2, 1.2)
 		do_teleport(M, src.target, 1) ///You will appear adjacent to the beacon
 
+/obj/warp_beacon/proc/startpack()
+	src.packable = 0
+	src.icon_state = "beaconpack"
+	SPAWN_DBG(14)
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:179]: actionbars or doafter, don't just use a spawn here

File: code/modules/transport/pods/warp_travel.dm [hunk 3]
```diff
@@ -144,4 +173,182 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+
+	New()
+		src.beaconid = rand(1000,9999)
+		src.name = "warp buoy unit [beaconid]"
+		..()
+
+	attackby(obj/item/W as obj, mob/user as mob)
+		if (istype(W, /obj/item/wrench) && !src.deploying)
+			for (var/turf/a in range(2,src))
+				if (!a.allows_vehicles)
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:207]:
```suggestion
			for (var/turf/T in range(2,src))
				if (!T.allows_vehicles)
```

File: code/modules/transport/pods/warp_travel.dm [hunk 4]
```diff
@@ -144,4 +173,182 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+					boutput(usr,"<span style=\"color:red\">The area surrounding the beacon isn't sufficiently navigable for vehicles.</span>")
+					return
+			if (isrestrictedz(src.z))
+				boutput(usr, "<span style=\"color:red\">The beacon can't connect to the warp network.</span>")
+				return
+//additional check, for if this thing gets used terribly somehow
+//			for (var/area/station/a in orange(1,src))
+//				boutput(user,"<span style=\"color:red\">Interference from nearby electrical fields is preventing this beacon's deployment.</span>")
+//				return
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:216]:
```suggestion
```

File: code/modules/transport/pods/warp_travel.dm [hunk 5]
```diff
@@ -144,4 +173,182 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+			src.beaconid = "[str]"
+			src.name = "warp buoy unit [beaconid]"
+			boutput(usr, "<span style=\"color:blue\">Designation updated to 'Buoy [str]'.</span>")
+		else
+			..()
+
+/obj/beacon_deployer/proc/deploybeacon()
+	src.icon_state = "beacondeploy"
+	src.anchored = 1
+	SPAWN_DBG(16)
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:239]: again, actionbar or do_after

File: code/modules/transport/pods/warp_travel.dm [hunk 6]
```diff
@@ -144,4 +173,182 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+	icon_state = "beacframe_1"
+	density = 1
+	var/state = 1
 
+	attackby(var/obj/item/I as obj, var/mob/user as mob)
+		switch(state)
+			if(1)
+				if (I.amount < 4)
+					boutput(user, "<span style=\"color:red\">You don't have enough rods to complete the stand (4 required).</span>")
+				else if (istype(I, /obj/item/rods))
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:260]: checking for rods should probably come before the amount check

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/modules/transport/pods/warp_travel.dm (hunk 6 above)
Comment (Kubius) [reply code/modules/transport/pods/warp_travel.dm]: Resolved with 693bfd0

File: code/modules/transport/pods/warp_travel.dm (hunk 2 above)
Comment (Kubius) [reply code/modules/transport/pods/warp_travel.dm]: Resolved with 693bfd0 - also affects the other one

File: code/modules/transport/pods/warp_travel.dm [hunk 7]
```diff
@@ -144,4 +173,179 @@ var/global/list/warp_beacons = list() //wow you should've made one for warp beac
...
+/obj/warp_beacon/proc/startpack()
+	src.packable = 0
+	src.icon_state = "beaconpack"
+	SPAWN_DBG(14) //wait until packing is complete
+		var/obj/beacon_deployer/packitup = new /obj/beacon_deployer(src.loc)
+		playsound(get_turf(src), "sound/machines/heater_off.ogg", 20, 1)
+		if(src.beaconid)
+			packitup.beaconid = src.beaconid
+			packitup.name = "warp buoy unit [beaconid]"
+		SPAWN_DBG(0)
```
Comment (Tarmunora) [inline code/modules/transport/pods/warp_travel.dm:185]: I doubt this spawn is needed...

Comment (Kubius) [reply code/modules/transport/pods/warp_travel.dm]: Ingame test confirmed this. Removed it as well as its counterpart in the deployment proc.

**Review by Tarmunora — APPROVED** [review]

**Pull request merged**